src/encoding.ts

```diff
@@ -0,0 +1,347 @@
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function normalizeEncodingArgs(data: any): [ string, number[] ] {
+	data = String(data);
+
+	let buffer = <number[]> [];
+
+	return [ data, buffer];
+}
+
+function validateDecodingArgs(data: any) {
+	return data == null ? false : true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [0, Infinity], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string (or string coerced object)
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export class Ascii {
+	/**
+	 * Encode a string or string coerced object into an Ascii byte buffer
+	 *
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): number[] {
+		let buffer: number[];
+
+		[ data, buffer ] = normalizeEncodingArgs(data);
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			buffer[i] = data.charCodeAt(i);
+		}
+
+		return buffer;
+	}
+
+	/**
+	 * Decode an Ascii encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: number[]): string {
```

Comment:
This would be more convenient as `decode(data: Uint8Array | Buffer | number[])` since Uint8Array and Buffer act the same as far as `decode` is concerned (so the same code path would work for all of them). You could use a type alias, like `type ByteBuffer = Uint8Array | Buffer | number[]`, giving `decode(data: ByteBuffer)`.